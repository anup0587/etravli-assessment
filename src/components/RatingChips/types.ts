export interface RatingChipsProps {
  ratings: Array<{
    Source: string;
    Value: string;
  }>;
}
