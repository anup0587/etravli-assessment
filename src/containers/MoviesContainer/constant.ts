export const tableHeading = ["Episode No.", "Title", "Rating", "Date"];

export const optionList = [
  {
    label: "Episode",
    value: "episode_id",
  },
  {
    label: "Year",
    value: "release_date",
  },
];
