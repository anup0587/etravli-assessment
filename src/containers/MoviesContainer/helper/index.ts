import { MovieState } from "../../../store/movies/type";

export const convertToPercentage = (ratingValue: string) => {
  if (ratingValue.includes("/")) {
    // ratingValue is in the format "x/y"
    const [numerator, denominator] = ratingValue.split("/");
    const percentage = (parseFloat(numerator) / parseFloat(denominator)) * 100;
    return percentage.toFixed(0) + "%";
  }

  if (ratingValue.includes("%")) {
    // ratingValue is already in percentage format
    return ratingValue;
  }
};

export const calculateAverageRating = (ratings: any) => {
  let total = 0;
  let count = 0;
  if (ratings.length>0) {
    for (const rating of ratings) {
      const value = rating.Value;

      // Check if the value is in "X/Y" format (e.g. "8.3/10")
      const matchXY = value.match(/^(\d+(\.\d+)?)\/(\d+)$/);
      if (matchXY) {
        const numerator = parseFloat(matchXY[1]);
        const denominator = parseFloat(matchXY[3]);
        total += (numerator / denominator) * 10;
        count++;
      }

      // Check if the value is in "%" format (e.g. "83%")
      const matchPercent = value.match(/^(\d+)%$/);
      if (matchPercent) {
        const percent = parseFloat(matchPercent[1]);
        total += percent / 10;
        count++;
      }
    }
  }

  // Calculate and return the average rating, rounded to one decimal place
  return count ? Math.round((total / count) * 10) / 10 : 0;
};

export const getFilteredMovie = (
  moviesList: MovieState[],
  searchInput: string,
  newOrderBy: string
) => {
  const lowerCaseSearchInput = searchInput.toLowerCase().trim();

  const data = [...moviesList].sort((a, b) => {
    switch (newOrderBy) {
      case "release_date":
        return (
          new Date(b.release_date).getTime() -
          new Date(a.release_date).getTime()
        );
      case "episode_id":
        return a.episode_id - b.episode_id;
      default:
        return 0;
    }
  });

  return lowerCaseSearchInput === ""
    ? data
    : data.filter((movie) =>
        movie.title.toLowerCase().includes(lowerCaseSearchInput)
      );
};
