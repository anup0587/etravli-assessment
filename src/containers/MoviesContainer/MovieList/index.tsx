import {
  Box,
  Grid,
  Rating,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@mui/material";
import React  from "react";
import { tableHeading } from "../constant";
import { MovieState } from "../../../store/movies/type";
import "./style.scss";
import clsx from "clsx";
import { MovieListProps } from "../type";

const MovieList = ({ list, onHandleClick, title }: MovieListProps) => (
  <Grid item xs={12} sm={6} md={7} className="movie_list_container">
    <Box className="movie_list_scroll_container">
      <TableContainer>
        <Table className="movie_list_table" stickyHeader> 
          <TableHead>
            <TableRow>
              {tableHeading.map((heading) => (
                <TableCell key={heading}>{heading}</TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {list &&
              list.map((movie: MovieState, index: number) => (
                <TableRow
                  hover
                  key={movie.episode_id}
                  onClick={() => onHandleClick(movie.title)}
                  className={clsx("movie_list_table_row", {
                    activeClass: movie.title === title,
                  })}
                >
                  <TableCell>{movie.episode_id}</TableCell>
                  <TableCell>{movie.title}</TableCell>
                  <TableCell>
                    <Rating
                      name="rating"
                      defaultValue={index + 2}
                      max={10}
                      disabled
                    />
                  </TableCell>
                  <TableCell>{movie.release_date}</TableCell>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </TableContainer>
    </Box>
  </Grid>
);

export default MovieList;
