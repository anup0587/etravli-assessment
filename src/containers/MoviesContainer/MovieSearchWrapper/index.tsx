import { Grid } from "@mui/material";
import React from "react";
import SearchFilter from "../../../components/SearchFilter";
import { MovieSearchWrapperPropd } from "../type";
import SortByOrder from "../../../components/SortByOrder";
import { optionList } from "../constant";
import "./style.scss";

const MovieSearchWrapper = ({
  inputValue,
  onChange,
  onClickItem,
  selectedValue,
}: MovieSearchWrapperPropd) => (
  <Grid item xs={12} container spacing={1} className="movie_search_wrapper">
    <Grid item xs={4} sm={1}>
      <SortByOrder
        optionList={optionList}
        onClickItem={onClickItem}
        selectedValue={selectedValue}
      />
    </Grid>
    <Grid item xs={8} sm={11}>
      <SearchFilter inputValue={inputValue} onChange={onChange} />
    </Grid>
  </Grid>
);
export default MovieSearchWrapper;
