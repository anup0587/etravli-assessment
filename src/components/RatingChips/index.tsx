import { Chip } from "@mui/material";
import React from "react";
import { convertToPercentage } from "../../containers/MoviesContainer/helper";
import "./style.scss";
import { RatingChipsProps } from "./types";

const RatingChips = ({ ratings }: RatingChipsProps) => (
  <>
    {ratings &&
      ratings.map((rating: any) => (
        <Chip
          key={rating.Source}
          label={`${rating.Source} -  ${convertToPercentage(rating.Value)}`}
          color="primary"
          variant="outlined"
          className="movie_rating_chip"
          data-testid="chip"
        />
      ))}
  </>
);

export default RatingChips;
