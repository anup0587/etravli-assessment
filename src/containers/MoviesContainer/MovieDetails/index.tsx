import { Grid, Box, Typography } from "@mui/material";
import React from "react";
import RatingChips from "../../../components/RatingChips";
import { calculateAverageRating } from "../helper";
import { MovieDetailsProps } from "../type";
import RatingStar from "../../../components/RatingStar";
import NoDataBox from "../../../components/NoDataBox";
import noMovie from "../../../assets/image/no_movie.jpg";

const MovieDetails = (props: MovieDetailsProps) => (
  <Grid item xs={12} sm={6} md={5} className="movie_details_container">
    {!props.details.Title && (
      <Grid container spacing={1} justifyContent="center">
        <NoDataBox image={noMovie} text="Please select the movie from list " />
      </Grid>
    )}
    {props.details.Title && (
      <Grid container spacing={1}>
        <Grid item xs={12}>
          <Typography variant="h5" gutterBottom>
            {props.details.Title}
          </Typography>
        </Grid>
        <Grid item xs={4} sm={4} md={4}>
          <Box mb={2}>
            <img
              src={props.details.Poster}
              alt="movie Image"
              width="100%"
              style={{ maxWidth: "300px" }}
            />
          </Box>
          <Typography variant="subtitle2" gutterBottom>
            Directed by: {props.details.Director}
          </Typography>
        </Grid>
        <Grid item xs={8} sm={8} md={8}>
          <Typography variant="body1" gutterBottom>
            {props.details.Description}
          </Typography>
        </Grid>
        <Grid item xs={12}>
          <RatingStar
            subTitle="Average Rating:"
            averageRating={calculateAverageRating(props.details?.Ratings)}
          />
        </Grid>
        <Grid item xs={12}>
          <RatingChips ratings={props.details.Ratings} />
        </Grid>
      </Grid>
    )}
  </Grid>
);

export default MovieDetails;
