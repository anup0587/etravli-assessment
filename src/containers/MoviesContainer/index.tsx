import { Container, Grid } from "@mui/material";
import React, { memo, useEffect, useState } from "react";
import { useAppDispatch, useAppSelector } from "../../store";
import {
  fetchMovieDetails,
  fetchMovies,
  setSearchByTitle,
} from "../../store/movies";
import Loader from "../../components/Loader";
import { getFilteredMovie } from "./helper";
import MovieSearchWrapper from "./MovieSearchWrapper";
import MovieList from "./MovieList";
import MovieDetails from "./MovieDetails";
import "./style.scss";

const MoviesContainer = () => {
  const dispatch = useAppDispatch();

  const { loading, movieDetail, movies, searchTerm } = useAppSelector(
    (state) => state.movieList
  );

  const [title, setTitle] = useState<string>("");
  const [orderBy, setOrderBy] = useState<string>("episode_id");

  const filteredMovies = getFilteredMovie(movies, searchTerm, orderBy);

  useEffect(() => {
    dispatch(fetchMovies());
  }, []);

  const onHandleClick = async (movieTitle: string) => {
    setTitle(movieTitle);
    await dispatch(fetchMovieDetails(movieTitle));
  };

  const onSearchByTitle = (event: React.ChangeEvent<HTMLInputElement>) => {
    const searchTerm = event.target.value;
    dispatch(setSearchByTitle(searchTerm));
  };

  const onClickItem = (value: string) => {
    setOrderBy(value);
  };
  
  return (<>
    {loading && <Loader loading={loading} />}
    {!loading && (<Container maxWidth="xl" className="movie_container">
           <Grid container spacing={1} mt={2}>
          <MovieSearchWrapper
            inputValue={searchTerm}
            onChange={onSearchByTitle}
            onClickItem={onClickItem}
            selectedValue={orderBy}
          />
          <Grid
            item
            xs={12}
            container
            spacing={1}
            className="movie_container_movie_list"
          >
            <MovieList
              list={filteredMovies}
              onHandleClick={onHandleClick}
              title={title}
            />
            <MovieDetails details={movieDetail} />
          </Grid>
        </Grid>
     
    </Container> )}
  </>
  
  );
};

export default memo(MoviesContainer);
