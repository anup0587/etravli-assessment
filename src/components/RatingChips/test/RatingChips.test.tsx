import { render } from "@testing-library/react";
import RatingChips from "..";
import { convertToPercentage } from "../../../containers/MoviesContainer/helper";
import "@testing-library/jest-dom/extend-expect";

import { describe, expect, it } from "vitest";

describe('RatingChips', () => {
    const ratings = [
      { Source: 'IMDb', Value: '8.0/10' },
      { Source: 'Rotten Tomatoes', Value: '75%' },
      { Source: 'Metacritic', Value: '68/100' },
    ];
  
    it('renders chips with correct labels', () => {
        const { getAllByTestId } = render(<RatingChips ratings={ratings} />);
        const chips = getAllByTestId('chip');
    
        expect(chips.length).toBe(ratings.length);
    
        ratings.forEach((rating, index) => {
          expect(chips[index]).toHaveTextContent(`${rating.Source} - ${convertToPercentage(rating.Value)}`);
        });
      });
  
    it('matches snapshot', () => {
      const { container } = render(<RatingChips ratings={ratings} />);
      expect(container).toMatchSnapshot();
    });
  });