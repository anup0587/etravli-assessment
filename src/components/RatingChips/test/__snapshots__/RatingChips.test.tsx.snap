// Vitest Snapshot v1

exports[`RatingChips > matches snapshot 1`] = `
<div>
  <div
    class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-outlinedPrimary movie_rating_chip css-196205-MuiChip-root"
    data-testid="chip"
  >
    <span
      class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
    >
      IMDb -  80%
    </span>
  </div>
  <div
    class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-outlinedPrimary movie_rating_chip css-196205-MuiChip-root"
    data-testid="chip"
  >
    <span
      class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
    >
      Rotten Tomatoes -  75%
    </span>
  </div>
  <div
    class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-outlinedPrimary movie_rating_chip css-196205-MuiChip-root"
    data-testid="chip"
  >
    <span
      class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
    >
      Metacritic -  68%
    </span>
  </div>
</div>
`;
